File: apps/api/app/schemas/meta_forms.py
"""Schemas for Meta lead form mapping."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.import_template import ColumnMappingItem, ColumnSuggestionResponse


class MetaFormSummary(BaseModel):
    """Summary of a Meta lead form and mapping status."""

    id: UUID
    form_external_id: str
    form_name: str
    page_id: str
    page_name: str | None
    mapping_status: str
    current_version_id: UUID | None
    mapping_version_id: UUID | None
    mapping_updated_at: datetime | None
    mapping_updated_by_name: str | None
    is_active: bool
    synced_at: datetime
    unconverted_leads: int
    total_leads: int
    last_lead_at: datetime | None


class MetaFormColumn(BaseModel):
    """Column metadata for a Meta form question."""

    key: str
    label: str | None
    question_type: str | None


class MetaFormMappingPreviewResponse(BaseModel):
    """Preview response for Meta form mapping."""

    form: MetaFormSummary
    columns: list[MetaFormColumn]
    column_suggestions: list[ColumnSuggestionResponse]
    sample_rows: list[dict]
    has_live_leads: bool
    available_fields: list[str]
    ai_available: bool
    mapping_rules: list[ColumnMappingItem] | None
    unknown_column_behavior: Literal["ignore", "metadata", "warn"] = "metadata"


class MetaFormMappingUpdateRequest(BaseModel):
    """Request to update a Meta form mapping."""

    column_mappings: list[ColumnMappingItem] = Field(default_factory=list)
    unknown_column_behavior: Literal["ignore", "metadata", "warn"] = "metadata"


class MetaFormMappingUpdateResponse(BaseModel):
    """Response after updating mapping."""

    success: bool
    mapping_status: str
    mapping_version_id: UUID | None
    message: str | None = None


class MetaFormSyncRequest(BaseModel):
    """Request to sync forms."""

    page_id: str | None = None

File: apps/web/lib/api/meta-forms.ts
import api from './index'
import type { ColumnMappingItem, ColumnSuggestion } from './import'

export interface MetaFormSummary {
    id: string
    form_external_id: string
    form_name: string
    page_id: string
    page_name: string | null
    mapping_status: string
    current_version_id: string | null
    mapping_version_id: string | null
    mapping_updated_at: string | null
    mapping_updated_by_name: string | null
    is_active: boolean
    synced_at: string
    unconverted_leads: number
    total_leads: number
    last_lead_at: string | null
}

export interface MetaFormColumn {
    key: string
    label: string | null
    question_type: string | null
}

export interface MetaFormMappingPreview {
    form: MetaFormSummary
    columns: MetaFormColumn[]
    column_suggestions: ColumnSuggestion[]
    sample_rows: Array<Record<string, string>>
    has_live_leads: boolean
    available_fields: string[]
    ai_available: boolean
    mapping_rules: ColumnMappingItem[] | null
    unknown_column_behavior: 'ignore' | 'metadata' | 'warn'
}

export interface MetaFormMappingUpdate {
    column_mappings: ColumnMappingItem[]
    unknown_column_behavior: 'ignore' | 'metadata' | 'warn'
}

export async function listMetaForms(): Promise<MetaFormSummary[]> {
    return api.get<MetaFormSummary[]>('/integrations/meta/forms')
}

export async function syncMetaForms(payload?: { page_id?: string }): Promise<{
    success: boolean
    message: string
}> {
    return api.post('/integrations/meta/forms/sync', payload ?? {})
}

export async function getMetaFormMapping(formId: string): Promise<MetaFormMappingPreview> {
    return api.get<MetaFormMappingPreview>(`/integrations/meta/forms/${formId}/mapping`)
}

export async function updateMetaFormMapping(
    formId: string,
    payload: MetaFormMappingUpdate
): Promise<{ success: boolean; message?: string }> {
    return api.put(`/integrations/meta/forms/${formId}/mapping`, payload)
}
